File: scala/nth-prime/src/test/scala/Prime.scala
object Prime {

  def isPrime(n: Int) = {
    val r = Math.floor(Math.sqrt(n)).toInt
    r < 5 || (5 until r + 1).forall { x => n % x != 0 }
  }

  lazy val possiblePrimes = Stream.from(6, 6).flatMap(x => List(x - 1, x + 1))

  lazy val primes = 2 #:: 3 #:: possiblePrimes.filter(isPrime)

  def nth(n: Int) = primes(n - 1)
}

File: java/nth-prime/src/main/java/Prime.java
import java.util.stream.IntStream;

public class Prime {
    public static int nth(int upTo) {
        return primes()
                .skip(upTo - 1)
                .findFirst()
                .getAsInt();
    }

    private static IntStream primes() {
        return IntStream.concat(IntStream.of(2, 3), possiblePrimes().filter(Prime::isPrime));
    }

    private static IntStream possiblePrimes() {
        return IntStream.iterate(6, i -> i + 6)
                .flatMap(i -> IntStream.of(i - 1, i + 1));
    }

    private static boolean isPrime(int n) {
        int r = (int) Math.floor(Math.sqrt(n));

        return r < 5 || IntStream.range(5, r + 1).allMatch(x -> n % x != 0);
    }
}